backend/core/rule/tree.py

```diff
@@ -50,15 +84,27 @@ def tree_into_dict(self):
 
 
 class TreeNode(object):
-    def __init__(self, data, func, namespace, metadata, *args):
+    def __init__(self, data, func, namespace, metadata, course_model,
+                 credit_info, hide_false=False, delete_used=False, *args):
```

Comment:
I think these new parameters should be tied into `dict`, since there are too many parameters.
And `metadata` parameter should be the last parameter, considering its meaning.
So I suggest like:
```python
def __init__(self, data, func, namespace, property, metadata, *args)
```
Also this can result shorter code in line 42~45.